kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/PromiseFactory.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PromiseFactory {
```

Comment:
I think the class, and all the methods could be package access.

Reply:
Also, some Javadoc would be nice.

Reply:
> I think the class, and all the methods could be package access. 

Hmm, I'm half way through doing so but starting to question if this is sufficiently generic behaviour that it should be public and available to filter authors. As I can see us wanting to use a timeout in say the `VaultKms` to ensure the http requests completed in a timely fashion.

But maybe that points at a different model, maybe thats pointing at a promise builder, something offering an API more like

```java
promiseBuilder()
.withTimeout(50, TimeUnit.Milliseconds)
.withLogger("filterX")
.withLogLevel(WARN)
.wrapping(incommingPromise)
.build();
```

Reply:
That's an interesting idea, but I think we should keep it out of the public API for now (i.e. this PR). I think we'll start to get a feeling for what we need as a result of implementing the AWS KMS integration and maybe when we start using the schema registry in other filters.

Reply:
Ok. I'll cut this to default visibility and we can revisit later.

Reply:
fixed in ce2ec763f491b1358ae5525177bb31054688b924